packages/protocol/contracts/identity/Random.sol

```diff
@@ -43,16 +61,51 @@ contract Random is IRandom {
     }
 
     // add entropy
-    _random = keccak256(abi.encodePacked(_random, randomness));
+    addRandomness(block.number, keccak256(abi.encodePacked(_random, randomness)));
 
     commitments[proposer] = newCommitment;
   }
 
+  function addRandomness(uint bn, bytes32 randomness) internal {
+    _random = randomness;
+    history[bn] = randomness;
+    if (historySize == 0) {
+      historyFirst = block.number;
+      historySize = 1;
+    }
+    else if (historySize > randomnessBlockRetentionWindow) {
+      delete history[historyFirst];
+      delete history[historyFirst+1];
+      historyFirst += 2;
+      historySize--;
+    }
+    else if (historySize == randomnessBlockRetentionWindow) {
+      delete history[historyFirst];
+      historyFirst++;
+    }
+    else /* historySize < randomnessBlockRetentionWindow) */ {
+      historySize++;
+    }
+  }
+
   function computeCommitment(bytes32 randomness) public pure returns (bytes32) {
     return keccak256(abi.encodePacked(randomness));
   }
 
   function random() external view returns (bytes32) {
     return _random;
   }
+
+  function getBlockRandomness(uint256 bn) external view returns (bytes32) {
+    return _getBlockRandomness(bn, block.number);
+  }
+
+  function _getBlockRandomness(uint256 bn, uint256 cur) internal view returns (bytes32) {
+    require(bn <= cur, "Cannot query randomness of future blocks");
+    require(bn > cur - historySize, "Cannot query randomness of old blocks");
+    if (randomnessBlockRetentionWindow < cur) {
```

Comment:
Consider combining this with the previous requirement since the error message is the same